src/v/storage/disk_log_impl.cc

```diff
@@ -509,6 +512,140 @@ segment_set disk_log_impl::find_sliding_range(
     return segs;
 }
 
+ss::future<> disk_log_impl::sliding_window_compact(
+  const compaction_config& cfg, std::optional<model::offset> new_start_offset) {
+    vlog(gclog.info, "[{}] running sliding window compaction", config().ntp());
+    auto segs = find_sliding_range(cfg, new_start_offset);
+    if (segs.empty()) {
+        vlog(gclog.info, "[{}] no more segments to compact", config().ntp());
+        co_return;
+    }
+    for (auto& seg : segs) {
+        auto result = co_await storage::internal::self_compact_segment(
+          seg,
+          _stm_manager,
+          cfg,
+          *_probe,
+          *_readers_cache,
+          _manager.resources(),
+          storage::internal::should_apply_delta_time_offset(_feature_table));
+
+        vlog(
+          gclog.debug,
+          "[{}] segment {} self compaction result: {}",
+          config().ntp(),
+          seg->reader().filename(),
+          result);
+    }
```

Comment:
was just a little surprised to see this here in that i didn't expect us to make an initial pass over the segments to self compact. what was the reasoning behind this?

Reply:
This is primarily here so we can reuse the reducer code:
- it's where aborted transactions are handled
- it's what writes compaction indexes, which lets us build the offset maps going through the compacted index instead of reading the entire segment (it's true self compaction makes a pass through the segment, but subsequent compactions may need to build the map again and it'd be nice to avoid rereading the segment in doing so)